### ISSUE TITLE
GoXLR - Dynamic Channel Count

### ISSUE THREAD

**Issue by FrostyCoolSlug** [alsa-project/alsa-ucm-conf#444]
The next GoXLR firmware release is going to add two new output channels for the full sized device, labeled a 'Stream Mix 2'. Outside of general usage, there's no way to determine whether these channels exist or not outside of the UCM2 definitions as the device is booting up.

Is there currently the facility to configure the devices virtual channels based on the number of channels reported by the hardware (so if there are 23 hardware channels, use the current configuration, if there are 25 hardware channels use an adjusted configuration), and if not, could you suggest the best way to handle this implementation?

Regards,
Craig McLure

**Comment by perexg** [comment]
Can we determine the firmware version using sysfs ? E.g. in `/sys/class/sound/card1/device/` tree (replace the sound card number with correct one).

**Comment by FrostyCoolSlug** [comment]
Apologies for the delay in getting back to you..

I've done some checking through sysfs, but I've not been able to find any differences between the various firmware versions, it doesn't appear anything from the AS Format Descriptors are stored anywhere in the device node, and other than the additional channels nothing in the USB description has changed.

We could possible get the firmware to 'hint' at something, but from our checks there's no formal (or at least, compliant) way to indicate a firmware version in the general USB descriptors.

**Comment by perexg** [comment]
Probably, we can parse `/proc/card#/stream0` contents to know the number of channels for this device to do proper configs. Could you attach the contents of this file?

**Comment by perexg** [comment]
Another way (and probably preferred) is to check the changes in USB descriptors accessible through sysfs.

Example path: `/sys/class/sound/card1/device/../descriptors` 

Replace `card1` with the proper card number. It would be nice to have samples (file contents in the binary form) for testing.

**Comment by FrostyCoolSlug** [comment]
I've created PR #520 to partially solve the recent pipewire issue. As mentioned above, the Mini will always have 21 channels  and the full device will have either 23 or 25 depending on firmware. Given that they're physically different devices, having different configs makes sense. 

The number of people running the beta firmware which grows the Full sized device to 25 channels is currently incredibly small, so in around 90% of cases this will solve the problem, and we can address the USB descriptors for the full device moving forward.

**Comment by perexg** [comment]
I tried to solve this in a bit different way. See PR #521. Could you, give a test, please?

**Comment by FrostyCoolSlug** [comment]
I've compiled and installed alsa-lib from git, and replaced 'String1' in the beta condition, and get the following error:

`ALSA lib ucm_subs.c:953:(uc_mgr_get_substituted_value) variable '${sys-card:[type=hex,pos=0x9c]device/../descriptors}' is not defined in this context!`

Without the String1 change, everything continues to work as before.

**Comment by FrostyCoolSlug** [comment]
Update, I sent `Syntax 8` in `USB-Audio.conf` and now the GoXLR correctly loads with 25 channels when on the beta firmware, 23 on the regular firmware, and 21 on the Mini.

So it looks like this is all working, thanks!

**Comment by FrostyCoolSlug** [comment]
Additionally with this solved I've also added the following to the bottom of `GoXLR-HiFi.conf`, which now also resolves the original problem:

```
If.mix2 {
        Condition {
                Type String
                String1 "${var:CaptureChannels}"
                String2 "25"
        }
        True.SectionDevice."Line6" {
                Comment "Broadcast Stream Mix 2"

                Value {
                        CapturePriority 200
                }

                Macro.pcm_split.SplitPCMDevice {
                        Name "goxlr_stereo_in"
                        Direction Capture
                        HWChannels "${var:CaptureChannels}"
                        Channels 2
                        Channel0 23
                        Channel1 24
                        ChannelPos0 FL
                        ChannelPos1 FR
                }
        }
}
```

Do you want me to add it as a separate PR once the existing one is merged in?

**Comment by perexg** [comment]
Yes, create another PR for the mix stream. I merged my changes, now.

**Comment by perexg** [comment]
I created PR #524 to not forget to apply 'Syntax 8' changes for beta firmware later.
